### PR TITLE
CA-966 Create v2 Workspace

### PR DESCRIPTION
**Pull request by gpolumbo-broad**
Because this PR isn't already big and complicated enough, take a look at hackathon PR too:  https://github.com/broadinstitute/rawls/pull/1318  this should extract and isolate all of the new logic and all of the old logic for create/clone workspace and hopefully make things a little more composable and easier to read in the future.

As a general approach to this set of changes, I tried to avoid proliferation of the pattern of making external API calls to other Terra services or to google from inside database transactions.  This is why there are some new contexts and work being done outside of `withNewWorkspaceContext` instead of completely within it.  

With the move of Service Perimeter logic to Create Workspace, I created:  https://broadworkbench.atlassian.net/browse/CA-1087 as a ticket to remind us to go back and clean up the old Service Perimeter logic for Billing Projects.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20201106_workspace_google_project_columns.xml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog logicalFilePath="dummy" xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+
+    <changeSet logicalFilePath="dummy" author="gpolumbo" id="GOOGLE_PROJECT_ON_WORKSPACE">
+        <renameColumn tableName="WORKSPACE" oldColumnName="GOOGLE_PROJECT" newColumnName="GOOGLE_PROJECT_ID" columnDataType="VARCHAR(254)" />
+        <addColumn tableName="WORKSPACE">
+            <column name="GOOGLE_PROJECT_NUMBER" type="VARCHAR(254)" >
```
Comment (dvoet) [inline core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20201106_workspace_google_project_columns.xml]: populate this from `RAWLS_BILLING_PROJECT`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 2]
```diff
@@ -19,18 +19,19 @@ import scala.language.postfixOps
...
+                            bucketName: String,
+                            workflowCollection: Option[String],
+                            createdDate: Timestamp,
+                            lastModified: Timestamp,
+                            createdBy: String,
+                            isLocked: Boolean,
+                            recordVersion: Long,
+                            workspaceVersion: String,
+                            googleProjectId: String,
+                            googleProjectNumber: String = "UNKNOWN") {
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: this should be `Option[String]` and I would not give it a default

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 3]
```diff
@@ -134,7 +137,8 @@ case class Workspace(
                       attributes: AttributeMap,
                       isLocked: Boolean,
                       workspaceVersion: WorkspaceVersion,
-                      googleProject: GoogleProjectId
+                      googleProjectId: GoogleProjectId,
+                      googleProjectNumber: GoogleProjectNumber
```
Comment (dvoet) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: this should be `Option[GoogleProjectNumber]`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 4]
```diff
@@ -1817,8 +1835,146 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+  }
+
+  // Since multiple Billing Projects can specify the same Service Perimeter, we need to:
+  // 1. Load all the Billing Projects that specify this servicePerimeterName
+  // 2. Load all the Workspaces in all of those Billing Projects
+  // 3. Collect all of the GoogleProjectNumbers from those Workspaces
+  // 4. Post that list to Google to overwrite the Service Perimeter's list of included Google Projects
+  // 5. Poll until ^ Google Operation is complete
+  // Throw exceptions if any of this goes awry
+  private def addGoogleProjectToPerimeter(googleProjectNumber: GoogleProjectNumber, servicePerimeterName: ServicePerimeterName): Future[Unit] = {
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: `googleProjectNumber` does not need to be in the parameter list because it `collectWorkspacesInPerimeter` should get it

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 5]
```diff
@@ -1817,8 +1835,146 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+    collectWorkspacesInPerimeter(servicePerimeterName).map { workspacesInPerimeter =>
+      val googleProjectNumbers: Seq[GoogleProjectNumber] = workspacesInPerimeter.map(_.googleProjectNumber) :+ googleProjectNumber
+      val projectNumberStrings = googleProjectNumbers.map(_.toString())
+
+      // Make the call to Google to overwrite the project.  Poll and wait for the Google Operation to complete
+      gcsDAO.accessContextManagerDAO.overwriteProjectsInServicePerimeter(servicePerimeterName, projectNumberStrings).map { operation =>
+        // This pollOperation is what we need to retry until the OperationStatus that gets returned is some terminal
+        // status
+        retryUntilSuccessOrTimeout(failureLogMessage = s"Google Operation to update Service Perimeter: ${servicePerimeterName} was not successful")(5 seconds, 50 seconds) { () =>
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1934]: 👍

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 6]
```diff
@@ -1817,8 +1835,146 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+    }
+  }
+
+  // In its own transaction, look up all of the Workspaces contained in Billing Projects that use the specified
+  // ServicePerimeterName
+  private def collectWorkspacesInPerimeter(servicePerimeterName: ServicePerimeterName): Future[Seq[Workspace]] = {
+    dataSource.inTransaction { dataAccess =>
+      for {
+        allProjectsWithPerimeter <- dataAccess.rawlsBillingProjectQuery.listProjectsWithServicePerimeterAndStatus(servicePerimeterName, CreationStatuses.all.toSeq: _*)
+        allWorkspacesInBillingProject <- dataAccess.workspaceQuery.listByNamespaces(allProjectsWithPerimeter.map(_.projectName))
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: this does 2 database queries, write a query to do it in 1

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 7]
```diff
@@ -1817,8 +1835,146 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+  }
+
+  // Gets a Google Project from the Resource Buffering Service (RBS) and sets it up to be usable by Rawls as the backing
+  // Google Project for a Workspace.  The specific entities in the Google Project (like Buckets or compute nodes or
+  // whatever) that are used by the Workspace will all get set up later after the Workspace is created in Rawls.  The
+  // project should NOT be added to any Service Perimeters yet, that needs to happen AFTER we persist the Workspace
+  // record.
+  // 1. Claim Project from RBS
+  // 2. Update Billing Account information on Google Project
+  // Returns a Future[(GoogleProjectId, GoogleProjectNumber)] of the project that we claimed from RBS
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: comment functions in real scaladoc

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 8]
```diff
@@ -1817,8 +1835,146 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+      case Some(servicePerimeterName) => addGoogleProjectToPerimeter(googleProjectNumber, servicePerimeterName)
+      case None => Future.successful()
+    }
+  }
+
+  // Some Service Perimeters may required that they have some additional non-Terra Google Projects that need to be in
+  // the perimeter for some other reason.  They provide those to us and we add them to the Rawls Config so that
+  // whenever we update the list of projects for a perimeter, these projects are always included
+  private def loadStaticProjectsForPerimeter(servicePerimeterName: ServicePerimeterName): Seq[String] = {
+    val staticProjectsConfig = ConfigFactory.load().getConfig("gcs.servicePerimeters.staticProjects")
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: config should be passed into the WorkspaceService via constructor args, not loaded on demand. This low form of dependency injection makes testing an reuse much easier.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 9]
```diff
@@ -1817,8 +1835,146 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+    * FALSE
+    */
+  def updateAndGetBillingAccountAccess(billingProject: RawlsBillingProject, parentSpan: Span = null): Future[Boolean] = {
+    val billingAccountName: RawlsBillingAccountName = billingProject.billingAccount.getOrElse(throw new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.InternalServerError, s"Billing Project ${billingProject.projectName.value} has no Billing Account associated with it")))
+    for {
+      hasAccess <- traceWithParent("checkBillingAccountIAM", parentSpan)(_ => gcsDAO.testDMBillingAccountAccess(billingAccountName.value))
+      updatedBillingProject = billingProject.copy(invalidBillingAccount = !hasAccess)
+      _ <- dataSource.inTransaction { dataAccess =>
+        traceDBIOWithParent("updateInvalidBillingAccountField", parentSpan)(_ => dataAccess.rawlsBillingProjectQuery.updateBillingProjects(Seq(updatedBillingProject)))
+      }
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: only do this if `invalidBillingAccount` has changed

**Comment by s-rubenstein** [comment]
Should this PR be against `v2_workspaces` instead of `develop`?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 10]
```diff
@@ -1817,8 +1834,178 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+    // We should never get here with a missing or invalid Billing Account, but we still need to get the value out of the
+    // Option, so we are being thorough
+    val billingAccount = billingProject.billingAccount match {
+      case Some(ba) if !billingProject.invalidBillingAccount => ba
+      case _ => throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"Billing Account is missing or invalid for Billing Project: ${billingProject}"))
+    }
+
+    for {
+      googleProjectId <- traceWithParent("getGoogleProjectFromRBS", span)(_ => getGoogleProjectFromRBS)
+      _ <- traceWithParent("updateBillingAccountForProject", span)(_ => gcsDAO.setBillingAccountForProject(googleProjectId, billingAccount))
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1870]: we also need to set google IAM on the project

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala [hunk 11]
```diff
@@ -22,28 +22,31 @@ import org.broadinstitute.dsde.rawls.webservice._
...
 import org.mockito.ArgumentMatchers._
 import org.mockito.Mockito
 import org.mockito.Mockito.{RETURNS_SMART_NULLS, verify}
+import org.mockito.Mockito.{RETURNS_SMART_NULLS, verify, when}
 import org.scalatest.prop.TableDrivenPropertyChecks
 
 import scala.concurrent.duration._
-import scala.concurrent.{Await, ExecutionContext}
+import scala.concurrent.{Await, ExecutionContext, Future}
+//import scala.concurrent.{Await, ExecutionContext, Future}
```
Comment (andy7i) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala]: drive-by comment for this commented out import
(just saw it while rebasing)

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 10 above)
Comment (gpolumbo-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1870]: We created a new ticket for taking care of this and any other permissions errors we might find:  https://broadworkbench.atlassian.net/browse/CA-1084

**Comment by gpolumbo-broad** [comment]
We need to make sure `cloneWorkspace` does the same as `createWorkspace`.  We need to make a call to `maybeUpdateGoogleProjectsInPerimeter` after we persist the Workspace. Unless we decide we want to encapsulate all of that inside `withNewWorkspaceContext`

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockBillingHttpGoogleServicesDAO.scala [hunk 12]
```diff
@@ -1,109 +0,0 @@
...
-    Future.successful(Seq(firecloudHasThisOne, firecloudDoesntHaveThisOne))
-  }
-
-  protected override def testDMBillingAccountAccess(billingAccountId: String): Future[Boolean] = {
-    billingAccountId match {
-      case "billingAccounts/firecloudHasThisOne" => Future.successful(true)
-      case "billingAccounts/firecloudDoesntHaveThisOne" => Future.successful(false)
-    }
-  }
-}
```
Comment (gpolumbo-broad) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockBillingHttpGoogleServicesDAO.scala]: Noticed this Mock was just never used

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 13]
```diff
@@ -1817,8 +1834,182 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+        retryUntilSuccessOrTimeout(failureLogMessage = s"Google Operation to update Service Perimeter: ${servicePerimeterName} was not successful")(5 seconds, 50 seconds) { () =>
+          gcsDAO.pollOperation(OperationId(GoogleApiTypes.AccessContextManagerApi, operation.getName)).map {
+            case OperationStatus(false, _) => Future.failed(new RawlsException(s"Google Operation to update Service Perimeter ${servicePerimeterName} is still in progress..."))
+            // TODO: If the operation to update the Service Perimeter failed, we need to consider the possibility that
+            // the list of Projects in the Perimeter may have been wiped or somehow modified in an undesirable way.  If
+            // this happened, it would be possible for Projects intended to be in the Perimeter are NOT in that
+            // Perimeter anymore, which is a problem.
+            case OperationStatus(true, errorMessage) if !errorMessage.isEmpty => Future.successful(throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError, s"Google Operation to update Service Perimeter ${servicePerimeterName} failed with message: ${errorMessage}")))
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1941]: Another thing to think about is the Projects per Perimeter limit.  What happens when we try to overwrite a Service Perimeter with 2001 projects if the limit is 2000 in the perimeter?  Will the whole request fail?  Will only the last record not be added to the perimeter?  Will exactly one project at random not be added to the perimeter?  What do we do about the project(s) that failed to get added?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala [hunk 14]
```diff
@@ -742,6 +742,16 @@ class HttpGoogleServicesDAO(
     jsonVersion.asYaml.spaces2
   }
 
+  override def setBillingAccountForProject(googleProjectId: GoogleProjectId, billingAccountName: RawlsBillingAccountName, billingEnabled: Boolean = true): Future[Unit] = {
+    implicit val service = GoogleInstrumentedService.Billing
+    val billingServiceAccountCredential = getBillingServiceAccountCredential
+    val billingManager = getCloudBillingManager(billingServiceAccountCredential)
+    val projectBillingInfo = new ProjectBillingInfo().setBillingAccountName(s"billingAccounts/${billingAccountName.value}").setBillingEnabled(billingEnabled)
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala]: where is `billingAccountName` coming from? I believe that we store the full `billingAccounts/...` string in the database, so if `billingAccountName` is loaded from the DB then I think we'd just want `s"${billingAccountName.value}"`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 15]
```diff
@@ -264,6 +278,15 @@ trait WorkspaceComponent {
       } yield workspace
     }
 
+    def getWorkspacesInPerimeter(servicePerimeterName: ServicePerimeterName): ReadAction[Seq[Workspace]] = {
+      val workspaces = for {
+        billingProject <- rawlsBillingProjectQuery.getProjectsWithPerimeterAndStatusQuery(servicePerimeterName, CreationStatuses.all.toSeq)
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:283]: What about billing projects with service perimeters and `CreationStatuses` of `Creating` or `Errored`? I suppose only `Ready` billing projects will actually have workspaces in them so it's probably not a concern...

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 13 above)
Comment (marctalbott) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1941]: Maybe we should be checking before and throwing an exception if the perimeter is full already? The 5000 total projects in any perimeter per GCP organization would complicate that check further

**Review by marctalbott — APPROVED** [review]
Looks good, couple questions and the one comment about `billingAccountName` that might be worth double checking

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala [hunk 16]
```diff
@@ -742,6 +742,17 @@ class HttpGoogleServicesDAO(
     jsonVersion.asYaml.spaces2
   }
 
+  override def setBillingAccountForProject(googleProjectId: GoogleProjectId, billingAccountName: RawlsBillingAccountName, billingEnabled: Boolean = true): Future[Unit] = {
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:745]: https://github.com/broadinstitute/rawls/pull/1315 has a more sophisticated implementation

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 17]
```diff
@@ -208,6 +218,10 @@ trait WorkspaceComponent {
       loadWorkspaces(findByIdsQuery(workspaceIds), attributeSpecs)
     }
 
+    def listByNamespaces(namespaceNames: Seq[RawlsBillingProjectName]): ReadAction[Seq[Workspace]] = {
+      loadWorkspaces(findByNamespacesQuery(namespaceNames))
+    }
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:223]: dead code

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 15 above)
Comment (dvoet) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:283]: creation status is going away soon anyway

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 18]
```diff
@@ -370,6 +393,10 @@ trait WorkspaceComponent {
       filter(rec => (rec.namespace === namespaceName.value))
     }
 
+    private def findByNamespacesQuery(namespaceNames: Seq[RawlsBillingProjectName]): WorkspaceQueryType = {
+      filter(_.namespace.inSetBind(namespaceNames.map(_.value)))
+    }
+
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:399]: dead code

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 13 above)
Comment (dvoet) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1941]: let it ride. I suspect the entire request will fail leaving the perimeter unchanged. The current published quota is 10k but Greg J. says it is actually 20k.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 19]
```diff
@@ -1818,8 +1835,182 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+    * 5. Poll until Google Operation to update the Service Perimeter gets to some terminal state
+    * Throw exceptions if any of this goes awry
+    *
+    * @param servicePerimeterName
+    * @return Future[Unit] indicating whether we succeeded to update the Service Perimeter
+    */
+  private def overwriteGoogleProjectsInPerimeter(servicePerimeterName: ServicePerimeterName): Future[Unit] = {
+    collectWorkspacesInPerimeter(servicePerimeterName).map { workspacesInPerimeter =>
+      val projectNumbers = workspacesInPerimeter.flatMap(_.googleProjectNumber) ++ loadStaticProjectsForPerimeter(servicePerimeterName)
+      val projectNumberStrings = projectNumbers.map(_.value)
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: should dedup `projectNumberStrings`, with v1 workspaces there could be duplicates

**Review by dvoet — APPROVED** [review]
a couple minor things

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 19 above)
Comment (gpolumbo-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: changed the signature for `AccessContextManagerDAO.overwriteProjectsInServicePerimeter` to require a `Set` of project numbers instead of a `Seq`
